File: Repositry/Implementations/CartService.cs
using ShopeForHomeAPI.Data;
using ShopeForHomeAPI.DTOs;
using ShopeForHomeAPI.Models;
using ShopeForHomeAPI.Repositry.Interfaces;

namespace ShopeForHomeAPI.Repositry.Implementations
{
    public class CartService : ICartServices
    {
        private readonly ApplicationDbContext _context;
        public CartService(ApplicationDbContext context)
        {
            _context = context;
        }
        // Add item to cart
        public IEnumerable<CartItem> GetCartItems(string userId)
        {
            return _context.CartItems
                .Where(ci => ci.UserId == userId)
                .ToList();
        }
        // Add item to cart
        public void AddToCart(CartItemDto item)
        {
            var existingItem = _context.CartItems
                .FirstOrDefault(ci => ci.UserId == item.userId && ci.ProductId == item.ProductId);

            if (existingItem != null)
            {
                existingItem.Quantity += item.Quantity;
            }
            else
            {
                var cartItem = new CartItem
                {
                    UserId = item.userId,
                    ProductId = item.ProductId,
                    Quantity = item.Quantity
                };
                _context.CartItems.Add(cartItem);
            }
            _context.SaveChanges();
        }

        // Update item quantity
        public void UpdateQuantity(int itemId, int quantity)
        {
            var cartItem = _context.CartItems.Find(itemId);
            if (cartItem != null)
            {
                cartItem.Quantity = quantity;
                _context.SaveChanges();
            }
        }
        // Remove item from cart
        public void RemoveItem(int itemId)
        {
            var cartItem = _context.CartItems.Find(itemId);
            if (cartItem != null)
            {
                _context.CartItems.Remove(cartItem);
                _context.SaveChanges();
            }
        }
        // Clear cart
        public void ClearCart(string userId)
        {
            var cartItems = _context.CartItems.Where(ci => ci.UserId == userId);
            _context.CartItems.RemoveRange(cartItems);
            _context.SaveChanges();
        }


    }
}

File: src/app/Services/cart-service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';


@Injectable({
  providedIn: 'root'
})
export class CartService {
  private apiUrl = 'https://localhost:7143/api/Cart';
  private cartItemsSubject = new BehaviorSubject<any[]>([]);
  cartItems$ = this.cartItemsSubject.asObservable();

  constructor(private http: HttpClient) { }

  loadCart(userId: string): void {
  this.http.get<any[]>(`${this.apiUrl}/${userId}`).subscribe(items => {
    this.cartItemsSubject.next(items);
  });
}

  isProductInCart(productId: number): boolean {
    return this.cartItemsSubject.getValue().some(item => item.productId === productId);
  }
  // 1. Get cart items for a user
  getCartItems(userId: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/${userId}`);
  }

  // 2. Add item to cart
  addToCart(item: any): Observable<string> {
    return this.http.post<string>(`${this.apiUrl}/add`, item);
  }

  // 3. Update item quantity
  updateQuantity(itemId: number, quantity: number): Observable<string> {
    const params = new HttpParams().set('quantity', quantity.toString());
    return this.http.put<string>(`${this.apiUrl}/update/${itemId}`, null, { params });
  }

  // 4. Remove item from cart
  removeItem(itemId: number): Observable<string> {
    return this.http.delete<string>(`${this.apiUrl}/remove/${itemId}`);
  }

  // 5. Clear cart for a user
  clearCart(userId: string): Observable<string> {
    return this.http.delete<string>(`${this.apiUrl}/clear/${userId}`);
  }

}
